src/pymor/reductors/mt.py

```diff
@@ -0,0 +1,156 @@
+from pymor.reductors.basic import LTIPGReductor
+from pymor.algorithms.gram_schmidt import gram_schmidt, gram_schmidt_biorth
+import numpy as np
+
+
+class MTReductor(BasicObject):
+    """Modal Truncation reductor.
+
+    See Section 9.2 in [A05]_.
+
+    Parameters
+    ----------
+    fom
+        The full-order |LTIModel| to reduce.
+    mu
+        |Parameter values|.
+    """
+
+    def __init__(self, fom, mu=None):
+        assert isinstance(fom, LTIModel)
+        if not isinstance(mu, Mu):
+            mu = fom.parameters.parse(mu)
+        assert fom.parameters.assert_compatible(mu)
+        self.fom = fom
+        self.mu = mu
+        self.V = None
+        self.W = None
+        self._pg_reductor = None
+
+    def reduce(self, r=None, decomposition='samdp', projection='orth',
+               symmetric=False, method_options=None):
+        """Modal Truncation.
+
+        Parameters
+        ----------
+        r
+            Order of the reduced model
+        decomposition
+            Algortihm to use for the decomposition:
+
+            -`'eigs'`: use scipy.linalg.eig algorithm
+            -`'samdp'`: find dominant poles using samdp algorithm
+        projection
+            Projection method:
+            - `'orth'`: projection matrices are orthogonalized with
+              respect to the Euclidean inner product
+            - `'biorth'`: projection matrices are biorthogolized with
+              respect to the E product
+        symmetric
+            True if Operator A is symmetric, False if not
+        method_options
+            optional dict with more options for the method used to calculate
+            eigenvalues and eigenvectors
+
+        Returns
+        -------
+        rom
+            Reduced-order model.
+        """
+        assert min(self.fom.B.source.dim, self.fom.C.range.dim) <= r < self.fom.order
+        assert projection in ('orth', 'biorth')
+        assert decomposition in ('eigs', 'samdp')
+        assert method_options is None or isinstance(method_options, dict)
+        if not method_options:
+            method_options = {}
+            method_options['which'] = 'LR'
+
+        self.V = self.fom.B.as_range_array().empty(reserve=r)
+        self.W = self.fom.C.as_source_array().empty(reserve=r)
+
+        if decomposition == 'eigs':
+            if self.fom.A.sparse:
+                A = self.fom.A.as_source_array().to_numpy()
+            else:
+                A = self.fom.A.matrix
+            which = method_options['which']
+            if symmetric:
+                poles, ev_r = eig(A, right=True)
+                if which == 'SM':
+                    idx = np.argsort(np.abs(poles))
+                elif which == 'LR':
+                    idx = np.argsort(-poles.real)
+                rev = ev_r[:, idx]
+                poles = poles[:r]
+                rev = NumpyVectorArray(rev[:, :r].T,
+                                       self.fom.B.as_range_array().space)
```

Comment:
It's nicer to write something like `self.fom.A.source.from_numpy(rev[:, :r].T)`.